demo/queued_agent_win.cc

```diff
@@ -0,0 +1,270 @@
+// Copyright 2022 The Chromium Authors.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include <windows.h>
+
+#include <fstream>
+#include <iostream>
+#include <queue>
+#include <string>
+
+#include "content_analysis/sdk/analysis_agent.h"
+
+using content_analysis::sdk::Agent;
+using content_analysis::sdk::Session;
+using content_analysis::sdk::ContentAnalysisRequest;
+using content_analysis::sdk::ContentAnalysisResponse;
+
+// This class maintains a list of outstanding content analysis requests to
+// process.  Each request is encapsulated in one Session.  Requests are handled
+// in FIFO order.
+class RequestQueue {
+ public:
+  RequestQueue() {
+    Init();
+  }
+  ~RequestQueue() {
+    Term();
+  }
+
+  // Push a new content analysis session into the queue.
+  void push(std::unique_ptr<Session> session) {
+    Enter();
+    sessions_.push(std::move(session));
+    WakeOne();
```

Comment:
swap this and the next line -- the lock doesn't need to be held when waking, and could end up forcing the woken thread to wait for the Leave.

Reply:
wow. nice thread. sounds like it's only an issue if you have consumer threads with differing priorities and you rely on priority-based scheduling. we're not doing that here, so i don't think it's significant. given that there's a reason to wake before unlocking, i'm fine leaving both implementations as you had them -- i like the symmetry. :-)

Reply:
Put back the way it was.